src/utils/p5Utils.js

```diff
@@ -33,75 +43,81 @@ class P5Util {
+   * Either it is in the p5Instance mode OR it is in the window
+   * @returns {boolean} if it is in p5
    */
   checkP5() {
     return !!this.p5Instance;
   }
 
   /**
-     * Convert a canvas to Blob
-     * @param {HTMLCanvasElement} inputCanvas 
-     * @returns {Blob} blob object
-     */
+   * Convert a canvas to a Blob object.
+   * @param {HTMLCanvasElement} inputCanvas
+   * @returns {Promise<Blob>}
+   */
   /* eslint class-methods-use-this: ["error", { "exceptMethods": ["getBlob"] }] */
   getBlob(inputCanvas) {
-    return new Promise((resolve) => {
+    return new Promise((resolve, reject) => {
       inputCanvas.toBlob((blob) => {
-        resolve(blob);
+        if (blob) {
+          resolve(blob);
+        } else {
+          reject(new Error('Canvas could not be converted to Blob.'));
+        }
       });
     });
   };
 
   /**
-     * Load image in async way.
-     * @param {String} url 
-     */
+   * Load a p5.Image from a URL in an async way.
+   * @param {string} url
+   * @return {Promise<p5.Image>}
+   */
   loadAsync(url) {
-    return new Promise((resolve) => {
+    return new Promise((resolve, reject) => {
       this.p5Instance.loadImage(url, (img) => {
         resolve(img);
+      }, () => {
+        reject(new Error(`Could not load image from url ${url}`));
       });
     });
   };
 
   /**
-     * convert raw bytes to blob object
-     * @param {Array} raws 
-     * @param {number} x 
-     * @param {number} y 
-     * @returns {Blob}
-     */
-  async rawToBlob(raws, x, y) {
+   * convert raw bytes to blob object
+   * @param {number[] | Uint8ClampedArray | ArrayLike<number>} raws
+   * @param {number} width
+   * @param {number} height
+   * @returns {Promise<Blob>}
+   */
+  async rawToBlob(raws, width, height) {
     const arr = Array.from(raws)
     const canvas = document.createElement('canvas'); // Consider using offScreenCanvas when it is ready?
     const ctx = canvas.getContext('2d');
 
-    canvas.width = x;
-    canvas.height = y;
+    canvas.width = width;
+    canvas.height = height;
 
-    const imgData = ctx.createImageData(x, y);
-    const { data } = imgData;
+    const imgData = ctx.createImageData(width, height);
+    const {data} = imgData;
 
-    for (let i = 0; i < x * y * 4; i += 1 ) data[i] = arr[i];
+    for (let i = 0; i < width * height * 4; i += 1) data[i] = arr[i];
     ctx.putImageData(imgData, 0, 0);
 
-    const blob = await this.getBlob(canvas);
-    return blob;
+    return this.getBlob(canvas);
   };
 
   /**
-     *  Conver Blob to P5.Image
-     * @param {Blob} blob 
-     * @param {Object} p5Img
-     */
+   * Convert Blob to P5.Image
+   * @param {Blob} blob
+   * Note: may want to reject instead of returning null.
+   * @returns {Promise<p5.Image | null>}
+   */
   async blobToP5Image(blob) {
-    if (this.checkP5()) {
-      const p5Img = await this.loadAsync(URL.createObjectURL(blob));
-      return p5Img;
+    if (this.checkP5() && typeof URL !== "undefined") {
```

Comment:
Hello! I'm just reviewing code for fun. Curious – can you not just say `if (this.checkP5() && URL) {...`? Is it necessary to use typeof in order to check that it's not undefined?

Reply:
Great question!  That particular check is part of a larger effort towards making sure that ml5 can run in a backend Node.js environment without error.  Here I am making sure that the global `URL` constructor exists before I try to use it.

If a variable such as `URL` is never declared anywhere (with `let`, `const`, `var`, `class`, etc.) then trying to access `if (URL)` is a fatal `ReferenceError`.  But checking `typeof URL !== "undefined"` is ok and lets you know that it's safe to use that variable.  We definitely have to do this when accessing `window`.

It might be that's it's not actually needed for `URL` because [Node does have its own `URL` class](https://nodejs.org/api/url.html#urlcreateobjecturlblob), which I didn't know until right now 😄

Reply:
Interesting! That makes sense to me. Appreciate the response – I've learned something new!